law_reader/common/summarization.py

```diff
@@ -0,0 +1,55 @@
+
+import os
+import openai
+import textwrap
+
+from langchain.chat_models import ChatOpenAI
+from langchain.chains import LLMChain
+from langchain.prompts import PromptTemplate
+from langchain.text_splitter import CharacterTextSplitter
+from langchain.text_splitter import RecursiveCharacterTextSplitter
+from langchain.prompts import PromptTemplate
+from langchain.docstore.document import Document
+from langchain.schema.document import Document
+from langchain.chains.summarize import load_summarize_chain
+from langchain.vectorstores import Chroma
+from langchain.embeddings import GPT4AllEmbeddings
+
+
+os.environ["OPENAI_API_KEY"] = ""
+llm = ChatOpenAI(temperature=0, model_name="gpt-3.5-turbo-16k")
+
+## Splits a long text into smaller text chunks, each containing up to 500 characters with a 100-character overlap.
+## It creates a list of Document objects, each containing one of the text chunks.
+def get_text_chunks_langchain(text):
+   text_splitter = CharacterTextSplitter(chunk_size=500, chunk_overlap=100)
+   docs = [Document(page_content=x) for x in text_splitter.split_text(text)]
+   return docs
+
+## Generates a summary for a given full text by breaking it into smaller chunks and then summarizing them.
+def get_gpt_summary(full_text):
+
+   ## Split the full text into smaller text chunks using a larger chunk size of 1000 characters because it produced better output
+   text_chunks = get_text_chunks_langchain(full_text)
+   text_splitter = RecursiveCharacterTextSplitter(chunk_size=1000, chunk_overlap=0)
+   all_splits = text_splitter.split_documents(text_chunks)
+
+   ## Create a vectorstore from the text chunks using Chroma and GPT-4 embeddings.
+   vectorstore = Chroma.from_documents(documents=all_splits, embedding=GPT4AllEmbeddings())
+
+   ## Define a prompt for summarization.
```

Comment:
APPROVE AFTER CHANGE: Provide more context as to why there are two prompts, and what each prompt is for: i.e. one for the LLMChain and one for the vectorstore

Reply:
Both the prompts have different use cases. The question asked in Chroma vector store similarity search allows the LLM chain to focus on the relevant documents for summarization instead of passing all documents to the chain. The retrieved documents are then used by the LLM chain to generate the summary using the prompt.

Reply:
I'd recommend including this information about each in the respective comments for both the prompt and the question variable -- that way, we have this information fully documented within the code itself.

Reply:
These comments are added to the code.